trieste/space.py

```diff
@@ -480,6 +502,175 @@ def __eq__(self, other: object) -> bool:
+        >>> rgb.to_tags(tf.constant([[1], [0], [2]]))
+        <tf.Tensor: shape=(3, 1), dtype=string, numpy=
+        array([[b'Green'],
+               [b'Red'],
+               [b'Blue']], dtype=object)>
+
+    """
+
+    def __init__(self, categories: int | Sequence[int] | Sequence[str] | Sequence[Sequence[str]]):
+        """
+        :param categories: Number of categories or category names. Can be an array for
+            multidimensional spaces.
+        """
+        if isinstance(categories, int) or any(isinstance(x, str) for x in categories):
+            categories = [categories]  # type: ignore[assignment]
+
+        if not isinstance(categories, Sequence) or not (
+            all(
+                isinstance(x, Sequence)
+                and not isinstance(x, str)
+                and all(isinstance(y, str) for y in x)
+                for x in categories
+            )
+            or all(isinstance(x, int) for x in categories)
+        ):
+            raise TypeError("Invalid category description: expected either numbers or names.")
+
+        elif any(isinstance(x, int) for x in categories):
+            category_lens: Sequence[int] = categories  # type: ignore[assignment]
+            if any(x <= 0 for x in category_lens):
+                raise ValueError("Numbers of categories must be positive")
+            tags = [tuple(f"{i}" for i in range(n)) for n in category_lens]
+        else:
+            category_names: Sequence[Sequence[str]] = categories  # type: ignore[assignment]
+            if any(len(ts) == 0 for ts in category_names):
+                raise ValueError("Category name lists cannot be empty")
+            tags = [tuple(ts) for ts in category_names]
+
+        self._tags = tags
+
+        ranges = [tf.range(len(ts)) for ts in tags]
+        meshgrid = tf.meshgrid(*ranges, indexing="ij")
+        points = (
+            tf.reshape(tf.stack(meshgrid, axis=-1), [-1, len(tags)]) if tags else tf.zeros([0, 0])
+        )
+
+        super().__init__(points)
+
+    def __repr__(self) -> str:
+        """"""
+        return f"CategoricalSearchSpace({self._tags!r})"
+
+    @property
+    def has_bounds(self) -> bool:
+        return False
+
+    @property
+    def lower(self) -> TensorType:
+        raise AttributeError("Categorical search spaces do not have numerical bounds")
+
+    @property
+    def upper(self) -> TensorType:
+        raise AttributeError("Categorical search spaces do not have numerical bounds")
+
+    @property
+    def tags(self) -> Sequence[Sequence[str]]:
+        """The tags of the categories."""
+        return self._tags
+
+    @property
+    def one_hot_encoder(self) -> EncoderFunction:
+        """A one-hot encoder for the numerical indices."""
+
+        def encoder(x: TensorType) -> TensorType:
+            if tf.rank(x) != 2:
+                raise ValueError(
+                    f"Invalid input for one-hot encoding: expected rank 2, got {tf.rank(x)}"
+                )
```

Comment:
Should we consider supporting an arbitrary number of batch dimensions, and not just 1? I think this will require flattening/unflattening them as `tf.keras.layers.CategoryEncoding` only supports 1D (for the one hot case).

Reply:
Good suggestion!